pennylane/shadows/classical_shadow.py

```diff
@@ -0,0 +1,201 @@
+    :math:`\ell`-local observables.
+
+    One can in principle also reconstruct the global state :math:`\sum_t \rho^{(t)}/T`, though it is not advisable nor practical for larger systems due to the exponential scaling
+
+    Args:
+        bitstrings (tensor): recorded measurement outcomes in random Pauli bases.
+        recipes (tensor): recorded measurement bases.
+    
+    .. seealso:: `PennyLane demo on Classical Shadows <https://pennylane.ai/qml/demos/tutorial_classical_shadows.html>`_, :func:`~.pennylane.classical_shadows`
+    
+    **Example**
+
+    We obtain the ``bitstrings`` and ``recipes`` via :func:`~.pennylane.classical_shadows`:
+
+    .. code-block:: python3
+
+        dev = qml.device("default.qubit", wires=range(2), shots=1000)
+        @qml.qnode(dev)
+        def qnode(x):
+            qml.Hadamard(0)
+            qml.CNOT((0,1))
+            qml.RX(x, wires=0)
+            return classical_shadow(wires=range(2))
+
+        bitstrings, recipes = qnode(0)
+        shadow = ClassicalShadow(bitstrings, recipes)
+    
+    After recording these ``T=1000`` quantum measurements, we can post-process the results to arbitrary local expectation values of Pauli strings.
+    For example, we can compute the expectation value of a Pauli string
+    
+    >>> shadow.expval(qml.PauliX(0) @ qml.PauliX(1), k=1)
+    (1.0079999999999998+0j)
+
+    or of a Hamiltonian
+    >>> H = qml.Hamiltonian([1., 1.], [qml.PauliZ(0)@qml.PauliZ(1), qml.PauliX(0)@qml.PauliX(1)])
+    >>> shadow.expval(H, k=1)
+    (2.2319999999999998+0j)
+
+    The parameter ``k`` is used to estimate the expectation values via the `median of means` algorithm. The case ``k=1`` corresponds to simply taking the mean
+    value over all local snapshots. ``k>1`` corresponds to splitting the ``T`` local snapshots into ``k`` equal parts, and taking the median of their individual means.
+    """
+
+    def __init__(self, bitstrings, recipes):
+        self.bitstrings = bitstrings
+        self.recipes = recipes
+
+        assert bitstrings.shape == recipes.shape
+        self.snapshots = len(bitstrings)
+
+        self.observables = [
+            qml.matrix(qml.PauliX(0)),
+            qml.matrix(qml.PauliY(0)),
+            qml.matrix(qml.PauliZ(0)),
+        ]
+
+    def local_snapshots(self, wires=None, snapshots=None):
+        r"""Compute the T x n x 2 x 2 local snapshots
+        i.e. compute :math:`3 U_i^\dagger |b_i \rangle \langle b_i| U_i - 1` for each qubit and each snapshot
+        """
+        bitstrings = self.bitstrings
+        recipes = self.recipes
+
+        if isinstance(snapshots, int):
+            pick_snapshots = np.random.choice(
+                np.arange(snapshots, dtype=np.int64), size=snapshots, replace=False
+            )
+            pick_snapshots = qml.math.convert_like(pick_snapshots, bitstrings)
+            bitstrings = qml.math.gather(bitstrings, pick_snapshots)
```

Comment:
Is gather the all-interface conformal way of doing clever indexing, i.e. `bitstrings = bitstrings[pick_snapshots]`?

Reply:
Yep, but it can't handle NumPy's fancy indexing to its full extent without some modification.

Reply:
Ah good to know! Is there any documentation on this? Its not in [docs/math](https://pennylane.readthedocs.io/en/stable/code/qml_math.html) and not in numpy

Reply:
Almost all of the functions in `single_dispatch.py` are missing from those docs. I'm not sure if this is something we're planning on adding, since for most functions the naming and behaviour across interfaces are consistent (for example `qml.abs`, `qml.conj`, etc.). Maybe @antalszava or @josh146 can shed some light here?

Reply:
Not sure about our original consideration of adding those functions to the docs, but we could start adding them to the docs at one point unless there are blockers. @eddddddy once Josh confirms this case and it proves to be a TODO, could you create a GitHub issue about it?

Reply:
Sure, I can create an issue and maybe start adding those.

Reply:
Ah, this is because `single_dispatch.py` magically loads and inserts into the namespace _every function_ available in NumPy/TF/Torch/JAX. Hence why we haven't documented this in the past, it would mean documentation the size of 4 libraries!

Reply:
Right, but how about documenting just the oddball functions that we had to manually implement (e.g. `gather`, `scatter`, `coerce`, etc.)? There are still quite a few but it seems to be manageable.